benches/end_to_end.rs

```diff
@@ -1,385 +1,400 @@
+
+    fn request_chunks(mut self, chunk: &'static [u8], cnt: usize) -> Self {
+        assert!(cnt > 0);
+        self.request_body = Some(chunk);
+        self.request_chunks = cnt;
+        self
+    }
+
+    fn response_body(mut self, body: &'static [u8]) -> Self {
+        self.response_body = body;
+        self
+    }
+
+    fn parallel(mut self, cnt: u32) -> Self {
+        assert!(cnt > 0, "parallel count must be larger than 0");
+        self.parallel_cnt = cnt;
+        self
+    }
+
+    fn bench(self, b: &mut test::Bencher) {
+        use std::sync::Arc;
+        let _ = pretty_env_logger::try_init();
+        // Create a runtime of current thread.
+        let rt = Arc::new(
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .expect("rt build"),
+        );
+        let exec = rt.clone();
+
+        let req_len = self.request_body.map(|b| b.len()).unwrap_or(0) as u64
+            * u64::max(1, self.request_chunks as u64);
+        let bytes_per_iter = (req_len + self.response_body.len() as u64) * self.parallel_cnt as u64;
+        b.bytes = bytes_per_iter;
+
+        let addr = spawn_server(&rt, &self);
+
+        let url: hyper::Uri = format!("http://{}/hello", addr).parse().unwrap();
+
+        let mut request_sender = prepare_client(&rt, &self, &addr);
+
+        let make_request = || {
+            let chunk_cnt = self.request_chunks;
+            let body = if chunk_cnt > 0 {
+                let (mut tx, body) = Body::channel();
+                let chunk = self
+                    .request_body
+                    .expect("request_chunks means request_body");
+                exec.spawn(async move {
+                    for _ in 0..chunk_cnt {
+                        tx.send_data(chunk.into()).await.expect("send_data");
+                    }
+                });
+                body
+            } else {
+                self.request_body
+                    .map(Body::from)
+                    .unwrap_or_else(Body::empty)
+            };
+            let mut req = Request::new(body);
+            *req.method_mut() = self.request_method.clone();
+            *req.uri_mut() = url.clone();
+            req
+        };
+
+        let mut send_request = |req: Request<Body>| {
+            let fut = request_sender.send_request(req);
+            async {
+                let res = fut.await.expect("client wait");
+                let mut body = res.into_body();
+                while let Some(_chunk) = body.data().await {}
+            }
+        };
+
+        if self.parallel_cnt == 1 {
+            b.iter(|| {
+                let req = make_request();
+                rt.block_on(send_request(req));
+            });
+        } else {
+            for _ in 0..self.parallel_cnt {
+                b.iter(|| {
+                    let req = make_request();
+                    rt.block_on(send_request(req));
+                });
+            }
```

Comment:
as you can see @seanmonstar we do this in sync way... If i try to run this async way, i will have an error because the connection is not ready...
What is the new approach for handling parallel request with the new "conn" things? I tried with poll_ready, to wait until the connection is ready before making the requests... 
What do you think is the best here ? Opening n connections to the server or using an approach like using the poll ready function ?

Reply:
also i see on src/client/conn/mod.rs in the example this line: 
`request_sender.ready().await?; `
wich i think is wrong because i can't find that method on the code 💢  (maybe im wrong)

Reply:
Yea, we don't want to use `block_on` there. Especially since the parallel option is meant to send them all at the same time.

- For HTTP/1, you'd need to make `parallel_cnt` amount of connections, and then spawn tasks looping on `ready()` and then `send_request()`.
- For HTTP/2, you can make just one connection, and send `parallel_cnt` requests at a time.

Reply:
Hey, already did the changes, as you can see in the pr, but when i run the benchs, have the following error, do you know how to solve it ? How is the best way to await the request sender until is ready to send a request ? 
![Screenshot 2022-08-13 at 11 29 36](https://user-images.githubusercontent.com/77611438/184478087-002809ef-9554-4d37-81ba-87f02e294449.png)

Reply:
Hey Robertoskr. When I was working on this I called the `ready` method from `tower::ServiceExt` on the request sender: `request_sender.ready().await`. From the comments above it seems like you tried this, but maybe you forgot to import the trait?

Reply:
Thanks! Just tried, that and works, but also tried with a tokio::sync::Lock and performs better, don't know if its safe thought, but the code works well.

Reply:
and for the trait: Yes, did't know that trait existed 😹